Navigation/include/ActsLUXEPipeline/LUXEMeasurementsCreator.hpp

```diff
@@ -0,0 +1,246 @@
+            Acts::SourceLink sl(simpleSL);
+            LUXEDataContainer::SimMeasurement SM{sl, scatteredParameters, trueVertex, static_cast<int>(sourceId)};
+            result.push_back(SM);
+        }
+    };
+/// Propagate the track create smeared measurements from local coordinates.
+
+    class MeasurementsCreator : public IAlgorithm {
+    public:
+        /// @brief The nested configuration struct
+        struct Config {
+            /// The output collection
+            std::string outputCollection = "Measurements";
+            // detector
+            std::shared_ptr<const Acts::Experimental::Detector> detector;
+            // variable bins magnetic field
+            std::shared_ptr<Acts::InterpolatedBFieldMap<LUXEMagneticField::vGrid>> BFieldPtr;
+        };
+
+        /// @brief Constructor
+        MeasurementsCreator(Config config, Acts::Logging::Level level)
+                : IAlgorithm("MeasurementsCreator", level),
+                  m_cfg(std::move(config)) {
+            m_outputMeasurements.initialize(m_cfg.outputCollection);
+        }
+        ~MeasurementsCreator() = default;
+
+        template <typename propagator_t, typename track_parameters_t>
+        LUXEDataContainer::SimMeasurements createMeasurements(const propagator_t& propagator,
+                                       const AlgorithmContext& ctx,
+                                       const track_parameters_t& trackParameters) const {
+            using namespace Acts::UnitLiterals;
+            using Actions = Acts::ActionList<LUXENavigator::MeasurementsCreatorAction>;
+            using Aborters = Acts::AbortList<Acts::EndOfWorldReached>;
+            // Set options for propagator
+            Acts::PropagatorOptions<Actions, Aborters> options(ctx.geoContext, ctx.magFieldContext);
+
+            auto& creator = options.actionList.get<LUXENavigator::MeasurementsCreatorAction>();
+            creator.sourceId = ctx.eventNumber;
+
+            // Launch and collect the measurements
+            auto result = propagator.propagate(trackParameters, options).value();
+            return std::move(result.template get<LUXEDataContainer::SimMeasurements>());
+        }
+
+        /// @brief The execute method
+        ProcessCode execute(const AlgorithmContext& ctx) const override {
+            using namespace Acts::UnitLiterals;
+
+            // Create IP covariance matrix from
+            // reasonable standard deviations
+            Acts::BoundVector ipStdDev;
+            ipStdDev[Acts::eBoundLoc0] = 100_um;
+            ipStdDev[Acts::eBoundLoc1] = 100_um;
+            ipStdDev[Acts::eBoundTime] = 25_ns;
+            ipStdDev[Acts::eBoundPhi] = 2_degree;
+            ipStdDev[Acts::eBoundTheta] = 2_degree;
+            ipStdDev[Acts::eBoundQOverP] = 1 / 100_GeV;
+            Acts::BoundSquareMatrix ipCov =
+                    ipStdDev.cwiseProduct(ipStdDev).asDiagonal();
+
+            // Create the measurements
+            std::random_device rd;
```

Comment:
What's with the random spectrum? Weren't we going with the flat energy distribution?

Reply:
Uniform energy distribution means that x_1 distribution will not be uniform, it will lack in events with high x_1 (low E). This distribution is similar to the actual p_z dist according to arka's files and gives a more or less flat distribution in position.